src/arrow_capsule.rs

```diff
@@ -0,0 +1,123 @@
+use std::ffi::c_void;
+
+use arrow::array::Array;
+use arrow::ffi::{FFI_ArrowArray, FFI_ArrowSchema};
+use arrow::record_batch::RecordBatch;
+use pyo3::exceptions::PyValueError;
+use pyo3::ffi::PyCapsule_New;
+use pyo3::prelude::*;
+use pyo3::types::PyCapsule;
+
+/// Creates a PyCapsule containing an ArrowSchema
+pub fn schema_to_pycapsule<'py>(
+    py: Python<'py>,
+    schema: &arrow::datatypes::Schema,
+) -> PyResult<Bound<'py, PyCapsule>> {
+    let schema_ptr = Box::into_raw(Box::new(FFI_ArrowSchema::try_from(schema).map_err(
+        |e| PyValueError::new_err(format!("Failed to convert schema to FFI format: {}", e)),
+    )?));
+
+    // SAFETY: `schema_ptr` points to a valid `FFI_ArrowSchema` allocated by `Box`,
+    // and we provide a proper release callback to clean it up
+    let capsule_ptr = unsafe {
+        PyCapsule_New(
+            schema_ptr as *mut c_void,
+            c"arrow_schema".as_ptr(),
+            Some(release_arrow_schema_capsule),
+        )
+    };
+
+    if capsule_ptr.is_null() {
+        // Clean up if capsule creation failed
+        // SAFETY: `schema_ptr` was created by `Box::into_raw` above, so we own it and can safely drop it
+        unsafe {
+            drop(Box::from_raw(schema_ptr));
+        }
+        return Err(PyValueError::new_err("Failed to create schema capsule"));
+    }
+
+    // SAFETY: `capsule_ptr` is a valid non-null `PyObject*` returned by `PyCapsule_New`
+    Ok(unsafe { Bound::from_owned_ptr(py, capsule_ptr) }.downcast_into()?)
+}
+
+/// Creates a PyCapsule containing an ArrowArray
+pub fn array_to_pycapsule<'py>(
+    py: Python<'py>,
+    array: &dyn Array,
+) -> PyResult<Bound<'py, PyCapsule>> {
+    let array_ptr = Box::into_raw(Box::new(FFI_ArrowArray::new(&array.to_data())));
+
+    // SAFETY: `array_ptr` points to a valid `FFI_ArrowArray` allocated by `Box`,
+    // and we provide a proper release callback to clean it up
+    let capsule_ptr = unsafe {
+        PyCapsule_New(
+            array_ptr as *mut c_void,
+            c"arrow_array".as_ptr(),
+            Some(release_arrow_array_capsule),
+        )
+    };
+
+    if capsule_ptr.is_null() {
+        // Clean up if capsule creation failed
+        // SAFETY: `array_ptr` was created by `Box::into_raw` above, so we own it and can safely drop it
+        unsafe {
+            drop(Box::from_raw(array_ptr));
+        }
+        return Err(PyValueError::new_err("Failed to create array capsule"));
+    }
+
+    // SAFETY: `capsule_ptr` is a valid non-null `PyObject*` returned by `PyCapsule_New`
+    Ok(unsafe { Bound::from_owned_ptr(py, capsule_ptr) }.downcast_into()?)
```

Comment:
Similarly you can simplify this and remove your own unsafe code by using `arrow-array::ffi`

https://github.com/kylebarron/arro3/blob/cb2453bf022d0d8704e56e81a324ab5a772e0247/pyo3-arrow/src/ffi/to_python/utils.rs#L53-L63